### ISSUE TITLE
[23.05] Build failure: python3.10-jupyter-server

### ISSUE THREAD

**Issue by jwrfoo** [NixOS/nixpkgs#267916]
### Steps To Reproduce

1. build python3.10-jupyter-server

### Build log

```
building '/nix/store/crxnqidkwkgxm9s7ahkhvilx1cisrki6-python3.10-jupyter-server-2.0.6.drv'...
error: builder for '/nix/store/crxnqidkwkgxm9s7ahkhvilx1cisrki6-python3.10-jupyter-server-2.0.6.drv' failed with exit code 1;
       last 10 log lines:
       >     pyfuncitem.obj(**testargs)
       >   File "/build/jupyter_server-2.0.6/tests/extension/test_launch.py", line 115, in test_token_file
       >     launch_instance(add_token=False)
       >   File "/build/jupyter_server-2.0.6/tests/extension/test_launch.py", line 83, in _run_in_subprocess
       >     wait_up(url, check=lambda: process.poll() is None)
       >   File "/build/jupyter_server-2.0.6/tests/extension/test_launch.py", line 40, in wait_up
       >     time.sleep(interval)
       >
       >  Timeout
       > /nix/store/gv2cl6qvvslz5h15vqd89f1rpvrdg5yc-stdenv-linux/setup: line 1604: pop_var_context: head of shell_variables not a function context
       For full logs, run 'nix log /nix/store/crxnqidkwkgxm9s7ahkhvilx1cisrki6-python3.10-jupyter-server-2.0.6.drv'.
error: 1 dependencies of derivation '/nix/store/xrsd0yj3cwww61kfwl5ypwjrkbl3qvvm-python3-3.10.13-env.drv' failed to build
error: 1 dependencies of derivation '/nix/store/wrq7h0rfcrd3ajv0l8jdr4ldjal7s0dr-man-paths.drv' failed to build
error: 1 dependencies of derivation '/nix/store/xplrny9gix16b24q0kfymx4pvgj6g894-system-path.drv' failed to build
error: 1 dependencies of derivation '/nix/store/4k4sdbxhkfwq2xj7y9xy8jr6jrmxjis7-nixos-system-cobalt-23.05.20231114.d4b5a67.drv' failed to build
```

Full build log here: https://paste.sr.ht/~jwr/605b829a9f0c61812bd1ff409708897029240ef6

### Additional context

I have the jupyter package in my system flake and it has worked fine up until recently.  I'm guessing jupyter-server is brought in as a dependency.  I have tried bisecting the issue and I believe the problem occurred between rev c41b26e688ff139e009e756a9b88c2760dfa8350 and fa9c4e560ee2b062c9ae38c75e499620f8cbf6a8.  I may be wrong on this as I'm still learning my way around nixos. 

### Notify maintainers
@fridh @elohmeier

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.0, NixOS, 23.05 (Stoat), 23.05.20231106.41de143`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.6`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by mcamp-ata** [comment]
`pkgs.jupyter`, `pkgs.python311Packages.jupyter_server` and `pkgs.python310Packages.jupyter_server` all seem to be broken for me.

**Comment by prokorpio** [comment]
`python310Packages.jupyterlab` is broken as well. I suppose all that are dependent on jupyter-server-2.0.6 has stopped working. 

Have you guys found any temporary workaround e.g. alternative jupyter-notebook package?

**Comment by mcamp-ata** [comment]
I believe it all worked around these commits `github:nixos/nixpkgs/5e47a07e9f2d7ed999f2c7943b0896f5f7321ca3`
`github:nixos/nixpkgs/85f1ba3e51676fa8cc604a3d863d729026a6b8eb`

as I went to an older version of my config and things build.. I grabbed those from the `flake.lock` think I got it correct.. waiting for it to build now..

**Comment by Gilk260** [comment]
`nbclassic`, `selenium`,.... all packages which require `jupyter_server` are broken

**Comment by prokorpio** [comment]
For those who are using `python3.withPackages` and need a workaround without using flakes, I found an alternative approach to create a python dev environment without relying much on `nixpkgs`, and simply use `pip` like in other distro's.

You can setup a `shell.nix` like this:

```Nix
{ pkgs ? import <nixpkgs> { } }:

let
  pythonEnv = pkgs.python3.withPackages(ps: [
    ps.pip
    ps.wheel
    ps.setuptools
  ]);

in
pkgs.mkShell {
  shellHook = ''
    export LD_LIBRARY_PATH=${pkgs.lib.makeLibraryPath [
      pkgs.stdenv.cc.cc
    ]}
  '';
  packages = [
    pythonEnv
  ];
}
```

You can then activate the nix shell to get python, then create and activate a python virtual environment via:
``` bash
nix-shell shell.nix
python -m venv venv
source venv/bin/activate
```

Finally, you can proceed to use pip as you would in other distro's. For example:
```bash
pip install jupyterlab
```

Additionally, you may use `direnv` so you can automatically enter the virtual environment when you `cd` into your dev directory. Simply create a `.envrc` with the following contents:
```bash
use_nix
source venv/bin/activate
```

Note that this is probably not very idiomatic from Nix perspective, although perhaps more stable.

### Some references
1. [Nixpkg manual on python development](https://nixos.org/manual/nixpkgs/stable/#how-to-consume-python-modules-using-pip-in-a-virtual-environment-like-i-am-used-to-on-other-operating-systems)
2. [LD_LIBRARY_PATH shellHook](https://discourse.nixos.org/t/nixos-with-poetry-installed-pandas-libstdc-so-6-cannot-open-shared-object-file/8442/9)

**Comment by risicle** [comment]
Fix is in staging-next-23.05 #267121

**Comment by tdolega** [comment]
> Fix is in staging-next-23.05 #267121

How to check if and when this will be merged into 23.05?

**Comment by alarsyo** [comment]
> How to check if and when this will be merged into 23.05?

You can check the status of the change on [this page](https://nixpk.gs/pr-tracker.html?pr=267121)

**Comment by natsukium** [comment]
close as complete
